### PR TITLE
Kubeadm types remove deps

### PR DESCRIPTION
**Pull request by mariantalla**
**What this PR does / why we need it**:

This PR removes some dependencies in `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/...`. This should
- Make the dependency graph a bit more unidirectional
- Aid in removing all dependencies to `k/k` (those should go away eventually, we should only depend on staging repos)

**Which issue(s) this PR fixes**

Addresses parts of https://github.com/kubernetes/kubeadm/issues/854

**Special notes for your reviewer**:

Number of packages `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/...` depends on:
- before: 167
- now: 50

Dependencies from `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/...` to other `kubeadm` packages:
- before:    
```sh
$ go list -f $'{{.ImportPath}}:{{range .Deps}}\n\t{{.}}{{end}}' 'k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/...' | ggrep 'kubeadm'
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm:
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/fuzzer:
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/scheme:
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2
        k8s.io/kubernetes/cmd/kubeadm/app/constants
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1:
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
        k8s.io/kubernetes/cmd/kubeadm/app/constants
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2:
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
        k8s.io/kubernetes/cmd/kubeadm/app/constants
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/validation:
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
        k8s.io/kubernetes/cmd/kubeadm/app/constants
        k8s.io/kubernetes/cmd/kubeadm/app/features
        k8s.io/kubernetes/cmd/kubeadm/app/util
        k8s.io/kubernetes/cmd/kubeadm/app/util/token
```
- after:  
```sh
$ go list -f $'{{.ImportPath}}:{{range .Deps}}\n\t{{.}}{{end}}' 'k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/...' | ggrep 'kubeadm'
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm:
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/constants:
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1:
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/constants
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2:
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
        k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/constants
```

<details>
<summary><a id="deps-per-pkg"></a>Dependencies per package (compare to <a href="https://github.com/kubernetes/kubeadm/issues/854#user-content-deps-per-pkg">before</a>)</summary>

```sh
$ go list -f $'{{.ImportPath}}:{{range .Deps}}\n\t{{.}}{{end}}' 'k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/...' \
  | ggrep 'k8s\.io' \
  | ggrep -v -P '\/vendor\/(?!k8s\.io)'
```

```
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm:
	k8s.io/kubernetes/pkg/apis/core
	k8s.io/kubernetes/pkg/apis/core/v1/helper/qos
	k8s.io/kubernetes/pkg/apis/scheduling
	k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig
	k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/v1beta1
	k8s.io/kubernetes/pkg/kubelet/qos
	k8s.io/kubernetes/pkg/kubelet/types
	k8s.io/kubernetes/pkg/master/ports
	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig
	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/v1alpha1
	k8s.io/kubernetes/pkg/util/pointer
	k8s.io/kubernetes/vendor/k8s.io/api/core/v1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/internalversion
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1beta1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/queryparams
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/fields
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/labels
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/schema
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/recognizer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/selection
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/types
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/errors
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/framer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/intstr
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/json
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/sets
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation/field
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/yaml
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/third_party/forked/golang/reflect
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/constants:
	k8s.io/kubernetes/pkg/apis/core
	k8s.io/kubernetes/pkg/registry/core/service/allocator
	k8s.io/kubernetes/pkg/registry/core/service/ipallocator
	k8s.io/kubernetes/vendor/k8s.io/api/core/v1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/internalversion
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1beta1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/queryparams
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/fields
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/labels
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/schema
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/recognizer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/selection
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/types
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/errors
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/framer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/intstr
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/json
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/sets
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation/field
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/yaml
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/third_party/forked/golang/reflect
	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/bootstrap/token/api
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1:
	k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
	k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/constants
	k8s.io/kubernetes/pkg/apis/core
	k8s.io/kubernetes/pkg/apis/core/v1/helper/qos
	k8s.io/kubernetes/pkg/apis/scheduling
	k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig
	k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/scheme
	k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/v1beta1
	k8s.io/kubernetes/pkg/kubelet/qos
	k8s.io/kubernetes/pkg/kubelet/types
	k8s.io/kubernetes/pkg/master/ports
	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig
	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/scheme
	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/v1alpha1
	k8s.io/kubernetes/pkg/registry/core/service/allocator
	k8s.io/kubernetes/pkg/registry/core/service/ipallocator
	k8s.io/kubernetes/pkg/util/pointer
	k8s.io/kubernetes/vendor/k8s.io/api/core/v1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/internalversion
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1beta1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/queryparams
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/fields
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/labels
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/schema
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/recognizer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/selection
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/types
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/errors
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/framer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/intstr
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/json
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/sets
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation/field
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/yaml
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/third_party/forked/golang/reflect
	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/bootstrap/token/api
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha2:
	k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
	k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/constants
	k8s.io/kubernetes/pkg/apis/core
	k8s.io/kubernetes/pkg/apis/core/v1/helper/qos
	k8s.io/kubernetes/pkg/apis/scheduling
	k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig
	k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/scheme
	k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/v1beta1
	k8s.io/kubernetes/pkg/kubelet/qos
	k8s.io/kubernetes/pkg/kubelet/types
	k8s.io/kubernetes/pkg/master/ports
	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig
	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/scheme
	k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/v1alpha1
	k8s.io/kubernetes/pkg/registry/core/service/allocator
	k8s.io/kubernetes/pkg/registry/core/service/ipallocator
	k8s.io/kubernetes/pkg/util/pointer
	k8s.io/kubernetes/vendor/k8s.io/api/core/v1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/internalversion
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1beta1
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/queryparams
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/fields
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/labels
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/schema
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/recognizer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/selection
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/types
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/errors
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/framer
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/intstr
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/json
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/sets
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation/field
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/yaml
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/third_party/forked/golang/reflect
	k8s.io/kubernetes/vendor/k8s.io/client-go/tools/bootstrap/token/api
```
</details>

<details>
<summary><a id="deps-all"></a>Dependencies combined and unified (compare to <a href="https://github.com/kubernetes/kubeadm/issues/854#user-content-deps-all">before</a>)</summary>

```sh
$ go list -f $'{{range .Deps}}\n{{.}}{{end}}' 'k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/...' \
  | ggrep 'k8s\.io' \
  | ggrep -v -P '\/vendor\/(?!k8s\.io)' \
  | sort \
  | uniq
```

```
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm
k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/constants
k8s.io/kubernetes/pkg/apis/core
k8s.io/kubernetes/pkg/apis/core/v1/helper/qos
k8s.io/kubernetes/pkg/apis/scheduling
k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig
k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/scheme
k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/v1beta1
k8s.io/kubernetes/pkg/kubelet/qos
k8s.io/kubernetes/pkg/kubelet/types
k8s.io/kubernetes/pkg/master/ports
k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig
k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/scheme
k8s.io/kubernetes/pkg/proxy/apis/kubeproxyconfig/v1alpha1
k8s.io/kubernetes/pkg/registry/core/service/allocator
k8s.io/kubernetes/pkg/registry/core/service/ipallocator
k8s.io/kubernetes/pkg/util/pointer
k8s.io/kubernetes/vendor/k8s.io/api/core/v1
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/api/resource
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/internalversion
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1beta1
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/conversion/queryparams
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/fields
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/labels
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/schema
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/json
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/protobuf
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/recognizer
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/serializer/versioning
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/selection
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/types
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/errors
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/framer
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/intstr
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/json
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/net
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/sets
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/validation/field
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/yaml
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch
k8s.io/kubernetes/vendor/k8s.io/apimachinery/third_party/forked/golang/reflect
k8s.io/kubernetes/vendor/k8s.io/client-go/tools/bootstrap/token/api
```
</details>

/sig cluster-lifecycle
/cc @hoegaarden 
/cc @luxas

### CODE REVIEW

**Comment by hoegaarden** [comment]
@dims @mariantalla is checking on the CLA already, that will hopefully be sorted out soon.

**Comment by mariantalla** [comment]
@dims Thanks, we made the change. We haven't worked with `.import-restrictions` in the past, feedback very welcome! We ran `hack/verify-import-boss.sh` before pushing, is there any other testing we should be doing?

**Comment by dims** [comment]
@mariantalla that should be it! i'll wait for others to chime in first about the rest of the changes, the .import-restrictions will depend on that feedback, so we can do that at the end.

**Comment by mariantalla** [comment]
@rosti - @hoegaarden and I unified constants again into `apis/kubeadm/constants/constants.go`. How does that look?

PTAL

**Comment by mariantalla** [comment]
It seems that `pull-kubernetes-kubemark-e2e-gce-big` is the only failing test; but it also fails at the same spot for recently merged PRs.

Does anyone have context on this? Do we need to worry about it? If yes, any advice on how to debug it?

cc @hoegaarden

**Comment by rosti** [comment]
@mariantalla Looks great now (although only a bit reordered from the original)! Thanks for the followup!
@hoegaarden `k8s.io/api/core/v1` is pulled in because of `MasterTaint` and `MasterToleration`. I think, that these may be moved to the places where they are used (they were not very spread through the source code). Also, they are not typical constants (more like const constructs) and I am not sure, but @luxas may move their usage to some place else in the future (due to the config split up).
So I think, that even with consts and vars merged, `k8s.io/api/core/v1` dependency can be avoided.

**Comment by hoegaarden** [comment]
Thanks @rosti -- let us know if there is anything you want us to change so it can be merged.

**Comment by rosti** [comment]
@hoegaarden @mariantalla Can you please rebase the PR and straighten up the history of constants.go (currently it shows a single blob of code being shifted further down the file). Also, this may be delayed a bit as everyone is on the mad dash to get feature PRs LGTM-ed prior to June 5th.

/assign @luxas 
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

**Comment by hoegaarden** [comment]
PTAL, thanks!

**Comment by hoegaarden** [comment]
**pull-kubernetes-bazel-test**:

> ERROR: /workspace/k8s.io/kubernetes/cmd/kubeadm/app/BUILD:27:1: target '//cmd/kubeadm/app/cmd:all-srcs' is not visible from target '//cmd/kubeadm/app:all-srcs'. Check the visibility declaration of the former target if you think the dependency is legitimate

I am not sure why this problem shows up or which commit triggers it. I can fix it by either

```diff
diff --git a/cmd/kubeadm/app/cmd/BUILD b/cmd/kubeadm/app/cmd/BUILD
index cd6a104eb9..c2a06d22ac 100644
--- a/cmd/kubeadm/app/cmd/BUILD
+++ b/cmd/kubeadm/app/cmd/BUILD
@@ -132,4 +132,5 @@ filegroup(
         "//cmd/kubeadm/app/cmd/util:all-srcs",
     ],
     tags = ["automanaged"],
+    visibility = ["//cmd/kubeadm/app:__pkg__"],
 )
```
or
```diff
diff --git a/cmd/kubeadm/app/cmd/BUILD b/cmd/kubeadm/app/cmd/BUILD
index cd6a104eb9..18d57194b2 100644
--- a/cmd/kubeadm/app/cmd/BUILD
+++ b/cmd/kubeadm/app/cmd/BUILD
@@ -132,4 +132,5 @@ filegroup(
         "//cmd/kubeadm/app/cmd/util:all-srcs",
     ],
     tags = ["automanaged"],
+    visibility = ["//visibility:public"],
 )
```
however, having not too much idea about bazel, I am not sure if this is a proper fix. However, if `k8s.io/kubernetes/cmd/kubeadm/app` has a dependency on `k8s.io/kubernetes/cmd/kubeadm/app/cmd` it might just be fine?

**pull-kubernetes-node-e2e**:

It seems to me, that this might be a flake and independent of our changes. I will trigger it once again
/test pull-kubernetes-node-e2e

File: cmd/kubeadm/app/phases/addons/dns/versions.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/*
```
Comment (chuckha) [inline cmd/kubeadm/app/phases/addons/dns/versions.go]: I think this PR needs a rebase, this file should not exist anymore

File: cmd/kubeadm/app/cmd/BUILD [hunk 2]
```diff
@@ -1,10 +1,4 @@
-package(default_visibility = ["//visibility:public"])
```
Comment (chuckha) [inline cmd/kubeadm/app/cmd/BUILD]: The problem with the bazel build (if i'm understanding bazel correctly) is that this line got deleted which makes the cmd:all-srcs target private. Since no default visibility is defined, the visibility will end up being private.

https://docs.bazel.build/versions/master/be/common-definitions.html#common.visibility

**Review by chuckha — COMMENTED** [review]
please rebase, there are a few files that shouldn't be in this PR (dns files at least). Also a comment about the bazel failure

File: cmd/kubeadm/app/apis/kubeadm/constants/BUILD [hunk 3]
```diff
@@ -1,15 +1,10 @@
...
-    "go_test",
-)
+load("@io_bazel_rules_go//go:def.bzl", "go_library", "go_test")
 
 go_library(
     name = "go_default_library",
     srcs = ["constants.go"],
-    importpath = "k8s.io/kubernetes/cmd/kubeadm/app/constants",
+    importpath = "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/constants",
+    visibility = ["//visibility:public"],
```
Comment (chuckha) [inline cmd/kubeadm/app/apis/kubeadm/constants/BUILD]: is this an automated change?

**Comment by hoegaarden** [comment]
@chuckha Thanks for your comments. The commits have been squashed, your comments have been addressed. PATL, thanks!

File: cmd/kubeadm/app/apis/kubeadm/constants/BUILD (hunk 3 above)
Comment (hoegaarden) [reply cmd/kubeadm/app/apis/kubeadm/constants/BUILD]: We didn't knowingly add that manually ... let me check ...

Comment (chuckha) [reply cmd/kubeadm/app/apis/kubeadm/constants/BUILD]: ah ok, no worries, was gong to suggest moving this to default visibility, but i prefer whatever the automated output is

Comment (hoegaarden) [reply cmd/kubeadm/app/apis/kubeadm/constants/BUILD]: Deleting `cmd/kubeadm/app/apis/kubeadm/constants/BUILD` and running `./hack/update-bazel.sh` generates the same file (some reordering, but no change of the `visibility` setting for `go_default_library`).
So, yes, automated change.

**Review by chuckha — COMMENTED** [review]
I like this change a lot, thank you for figuring this out! 

I noticed that the import identifier changes a bunch throughout and it's not consistent (which it already was not consistent) which makes for a bigger diff than strictly necessary.

I propose changing the import identifier is out of scope for what this PR is doing. Would you be willing to keep the identifiers the same? (kubeadmconstants vs constants)

File: cmd/kubeadm/app/phases/upgrade/staticpods.go [hunk 4]
```diff
@@ -84,11 +84,11 @@ func NewKubeStaticPodPathManagerUsingTempDirs(realManifestDir string, saveManife
 	if err != nil {
 		return nil, err
 	}
-	backupManifestsDir, err := constants.CreateTimestampDirForKubeadm("kubeadm-backup-manifests")
```
Comment (chuckha) [inline cmd/kubeadm/app/phases/upgrade/staticpods.go]: why did this function change?

**Comment by timothysc** [comment]
shifting to 1.12, we're currently in code freeze for 1.11

File: cmd/kubeadm/app/phases/upgrade/staticpods.go (hunk 4 above)
Comment (hoegaarden) [reply cmd/kubeadm/app/phases/upgrade/staticpods.go]: - This is probably a mistake I made while resolving conflicts on a rebase
- The fact that all test came back green suggests that this is still ok
- I assume, that there is some reason why this change was made and something/someone is expecting a timestamped directory - if so, I think there should be some sort of test around that then
- I will try to add a test for that (if meaningful and reasonably possible -- nee to check) to guard against that

Comment (chuckha) [reply cmd/kubeadm/app/phases/upgrade/staticpods.go]: I'm ok with omitting a test, that is out of scope for this PR to me, if you remove this change then it all looks good to me!

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 5]
```diff
@@ -26,7 +26,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	clientsetfake "k8s.io/client-go/kubernetes/fake"
 	core "k8s.io/client-go/testing"
-	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
```
Comment (chuckha) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go]: I think one more import rename snuck through

**Review by chuckha — COMMENTED** [review]
Two comments and then this looks good to me

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go (hunk 5 above)
Comment (hoegaarden) [reply cmd/kubeadm/app/phases/addons/dns/dns_test.go]: Has been fixed.

**Review by luxas — COMMENTED** [review]
(I don't have time to review this now line-by-line, as we're in bugfix/release mode for v1.11, sorry)

Thanks for this PR and initiative!
However, I'm a bit sceptical to some of the changes, e.g. the move of the `scheme` and `fuzzer` packages.
We should definitely do a better split between defaults in the kubeadm API types and the `constants` package, but that's a way broader discussion and needs some kind of intentional design planning.

More than just going ahead and moving packages, I want some kind of motivation for doing so. I guess you want to reduce the imports from the internal/external API types, but that requires other, more general changes like not vendoring k8s.io/kubernetes/... componentconfig types like we do now. For that to happen we need to talk to SIG API Machinery and maybe SIG Architecture.

TL;DR; before merging this I want to see some kind of end-to-end evidence with the desired end goal we all (including the rest of the k8s community) agrees on.

**Comment by mariantalla** [comment]
We'll investigate the API machinery a bit more to understand required changes a bit better. Next steps for the next iteration of this effort: write up a proposal that we present to the sig before starting work on a similar PR.

cc @hoegaarden
